### PR TITLE
[locator] autocomplete + allow restricting search to a single field in active layer filter

### PR DESCRIPTION
**Pull request by 3nids**
### enable auto completion in locator
a locator filter can now return a completion list while preparing the search
the line edit will use the first matching completion and display it as light grey text
the completion can be triggered by pressing Tab key

### active layer locator filter field name restriction
the filter can now restrict the search to a specific field by typing @+field_name
auto-completion is provided for the field names

### demo
locator

(included in this PR the commit from #38628 )

### CODE REVIEW

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 1]
```diff
@@ -526,6 +592,38 @@ void QgsAllLayersFeaturesLocatorFilter::triggerResultFromAction( const QgsLocato
...
+  std::unique_ptr<QDialog> dlg( new QDialog( parent ) );
+  dlg->restoreGeometry( settings.value( QStringLiteral( "Windows/%1/geometry" ).arg( key ) ).toByteArray() );
+  dlg->setWindowTitle( "All layers features locator filter" );
+  QFormLayout *formLayout = new QFormLayout;
+  QSpinBox *globalLimitSpinBox = new QSpinBox( dlg.get() );
+  globalLimitSpinBox->setValue( settings.value( QStringLiteral( "%1/limit_global" ).arg( key ), 15, QgsSettings::App ).toInt() );
+  globalLimitSpinBox->setMinimum( 1 );
+  globalLimitSpinBox->setMaximum( 200 );
+  formLayout->addRow( tr( "&Maximum number of results:" ), globalLimitSpinBox );
+  QSpinBox *parLayerLimitSpinBox = new QSpinBox( dlg.get() );
```
Comment (uclaros) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:608]: Should be `perLayerLimitSpinBox `?

File: src/core/locator/qgslocator.h [hunk 2]
```diff
@@ -171,6 +191,8 @@ class CORE_EXPORT QgsLocator : public QObject
     QList< QgsLocatorFilter * > mFilters;
     QList< QThread * > mActiveThreads;
 
+    QStringList mAutocompleList;
```
Comment (uclaros) [inline src/core/locator/qgslocator.h:194]: Should be something like `mAutocompletionList`, or is it intentional?

File: src/app/locator/qgsinbuiltlocatorfilters.cpp (hunk 1 above)
Comment (3nids) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp]: fixed, thanks

File: src/core/locator/qgslocator.h (hunk 2 above)
Comment (3nids) [reply src/core/locator/qgslocator.h]: fixed, thanks

File: python/core/auto_generated/locator/qgslocatorfilter.sip.in [hunk 3]
```diff
@@ -163,12 +163,13 @@ results from this filter.
...
 %End
 
-    virtual void prepare( const QString &string, const QgsLocatorContext &context );
+    virtual QStringList prepare( const QString &string, const QgsLocatorContext &context );
 %Docstring
 Prepares the filter instance for an upcoming search for the specified ``string``. This method is always called
 from the main thread, and individual filter subclasses should perform whatever
 tasks are required in order to allow a subsequent search to safely execute
 on a background thread.
+The method return an autocompletion list
```
Comment (m-kuhn) [inline python/core/auto_generated/locator/qgslocatorfilter.sip.in:172]:
```suggestion
The method returns an autocompletion list
```

File: src/gui/locator/qgslocatorwidget.h [hunk 4]
```diff
@@ -171,6 +174,30 @@ class GUI_EXPORT QgsLocatorResultsView : public QTreeView
...
+ * Custom line edit to handle completion within the line edit as a light gray text
+ * \since QGIS 3.16
+ */
+class QgsLocatorLineEdit : public QgsFilterLineEdit
+{
+    Q_OBJECT
+  public:
+    explicit QgsLocatorLineEdit( QgsLocatorWidget *locator, QWidget *parent = nullptr );
+
+    //! Perform completion and returns true if successful
```
Comment (m-kuhn) [inline src/gui/locator/qgslocatorwidget.h:190]:
```suggestion
    //! Performs completion and returns true if successful
```

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 5]
```diff
@@ -290,12 +321,16 @@ void QgsActiveLayerFeaturesLocatorFilter::prepare( const QString &string, const
...
+
+  return completionList;
 }
 
 void QgsActiveLayerFeaturesLocatorFilter::fetchResults( const QString &string, const QgsLocatorContext &, QgsFeedback *feedback )
 {
   int found = 0;
   QgsFeature f;
+  QString searchString = string;
+  fieldRestriction( searchString );
```
Comment (m-kuhn) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:333]: Should that use the return value somewhere?

File: src/core/locator/qgslocator.h [hunk 6]
```diff
@@ -153,6 +160,19 @@ class CORE_EXPORT QgsLocator : public QObject
      */
     void foundResult( const QgsLocatorResult &result );
 
+    /**
+     * Emitted when locator has begun a search, before actually preparing it.
+     * \since QGIS 3.16
+     */
+    void searchBegan();
```
Comment (m-kuhn) [inline src/core/locator/qgslocator.h:167]: What is the purpose of this? I don't see it currently used in this PR.

File: src/app/locator/qgsinbuiltlocatorfilters.h [hunk 7]
```diff
@@ -102,18 +102,27 @@ class APP_EXPORT QgsActiveLayerFeaturesLocatorFilter : public QgsLocatorFilter
...
     void fetchResults( const QString &string, const QgsLocatorContext &context, QgsFeedback *feedback ) override;
     void triggerResult( const QgsLocatorResult &result ) override;
+    bool hasConfigWidget() const override {return true;}
+    void openConfigWidget( QWidget *parent ) override;
 
   private:
 
+    /**
+     * Returns the field restriction if defined (starting with @)
+     * The \a searchString is modified acccordingly by removing the field restriction
```
Comment (m-kuhn) [inline src/app/locator/qgsinbuiltlocatorfilters.h:115]:
```suggestion
     * The \a searchString is modified accordingly by removing the field restriction
```

File: src/app/locator/qgsinbuiltlocatorfilters.cpp (hunk 5 above)
Comment (3nids) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:333]: nope, the return value is only used in the other call.

File: src/core/locator/qgslocator.h (hunk 6 above)
Comment (3nids) [reply src/core/locator/qgslocator.h:167]: I thought it would be required to update the completion text in the line edit but apparently not.
Removed

**Comment by 3nids** [comment]
Merging for broader testing, tests will follow

**Pull request merged**